File: Week5/Mini Projects/JournalC/JournalC/EntryController.m
//
//  EntryController.m
//  Journal-C
//
//  Created by Caleb Hicks on 12/1/15.
//  Copyright © 2015 DevMountain. All rights reserved.
//

#import "EntryController.h"

static NSString * const EntriesKey = @"entries";


@implementation EntryController

- (void)addEntry:(Entry *)entry {
    
    [self.entries addObject:entry];
    [self saveToPersistentStorage];
}

- (void)removeEntry:(Entry *)entry {
    
    [self.entries removeObject:entry];
    [self saveToPersistentStorage];
}

- (void)saveToPersistentStorage {
    
    NSMutableArray *entryDictionaries = [NSMutableArray new];
    
    for (Entry *entry in self.entries) {
        
        [entryDictionaries addObject:entry.dictionaryCopy];
    }
    
    [[NSUserDefaults standardUserDefaults] setObject:entryDictionaries forKey:EntriesKey];
}

- (void)loadFromPersistentStorage {
    
    NSMutableArray *entries = [NSMutableArray new];
    
    NSArray *entryDictionaries = [[NSUserDefaults standardUserDefaults] objectForKey:EntriesKey];
    
    for (NSDictionary *dictionary in entryDictionaries) {
        
        Entry *entry = [[Entry alloc] initWithDictionary:dictionary];
        [entries addObject:entry];
    }
    
    self.entries = entries;
}

+ (EntryController *)sharedController {
    static EntryController *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [EntryController new];
        [sharedInstance loadFromPersistentStorage];
    });
    return sharedInstance;
}

@end

File: Week1/Journal3/Journal3/EntryController.swift
//
//  EntryController.swift
//  Journal3
//
//  Created by Kaleo Kim on 2/7/16.
//  Copyright © 2016 Kaleo Kim. All rights reserved.
//

import Foundation

class EntryController {
    
    private let entryKey = "entries"
    
    static let sharedInstance = EntryController()
    
    var entries: [Entry]
    
    init() {
        entries = []
    }
    
    func addEntry(entry: Entry) {
        entries.append(entry)
    }
    
    func removeEntry(entry: Entry) {
        if let entryIndex = entries.indexOf(entry) {
            entries.removeAtIndex(entryIndex)
        }
    }
    
    func saveToPersistentStorage() {
        let entryDictionay = self.entries.map({$0.dictionaryCopy()})
        
        NSUserDefaults.standardUserDefaults().setObject(entryDictionay, forKey: entryKey)
    }
    
    func loadFromPersistentStorage() {
        
        let entryDictionaryFromDefaults = NSUserDefaults.standardUserDefaults().objectForKey(entryKey) as? [Dictionary<String, AnyObject>]
        
        if let entryDictionary = entryDictionaryFromDefaults {
            self.entries = entryDictionary.map({Entry(dictionary: $0)!})
        }
    }
    
}